src/controllers/legacycontrollermapping.cpp

```diff
@@ -0,0 +1,66 @@
+#include "controllers/legacycontrollermapping.h"
+
+void LegacyControllerMapping::restoreSettings(const QFileInfo& mappingFile,
+        UserSettingsPointer pConfig,
+        const QString& controllerName) {
+    QString controllerKey = QString(CONTROLLER_SETTINGS_PREFERENCE_GROUP_KEY)
+                                    .arg(controllerName, mappingFile.absoluteFilePath());
+
+    auto availableSettings = getSettings();
+    QList<ConfigKey> definedSettings = pConfig->getKeysWithGroup(controllerKey);
+
+    QList<QString> availableSettingKeys;
+    for (const auto& pSetting : qAsConst(availableSettings)) {
+        availableSettingKeys.append(pSetting->variableName());
+    }
+
+    bool ok;
+    for (const auto& key : definedSettings) {
+        if (!availableSettingKeys.contains(key.item)) {
+            qDebug() << "The setting" << key.item
+                     << "does not seem to exist in the mapping" << mappingFile.absoluteFilePath()
+                     << ". It may be invalid or may have been removed.";
+            pConfig->remove(key);
+            continue;
+        }
+        const auto& pSetting = availableSettings.at(availableSettingKeys.indexOf(key.item));
+        QString value = pConfig->getValueString(key);
+        if (!pSetting->valid()) {
+            qWarning() << "The setting" << pSetting->variableName()
+                       << "for the mapping" << mappingFile.absoluteFilePath()
+                       << "appears to be invalid. Its saved value won't be restored.";
+        }
+        pSetting->parse(value, &ok);
+        if (!ok || !pSetting->valid()) {
+            qWarning() << "The setting" << pSetting->variableName()
+                       << "for the mapping" << mappingFile.absoluteFilePath()
+                       << "could not be restore. Removing and resetting the setting default value.";
+            pConfig->remove(key);
+            pSetting->reset();
+        }
+    }
+}
+
+void LegacyControllerMapping::saveSettings(const QFileInfo& mappingFile,
+        UserSettingsPointer pConfig,
+        const QString& controllerName) {
+    QString controllerKey = QString(CONTROLLER_SETTINGS_PREFERENCE_GROUP_KEY)
+                                    .arg(controllerName, mappingFile.absoluteFilePath());
+    for (auto setting : getSettings()) {
+        if (!setting->isDirty()) {
+            continue;
+        }
+        setting->save();
+        if (!setting->valid()) {
+            qWarning() << "Setting" << setting->variableName()
+                       << "for controller" << controllerName
+                       << "is invalid. Its value will not be saved.";
+            continue;
+        }
+        if (setting->isDefault()) {
+            pConfig->remove(ConfigKey(controllerKey, setting->variableName()));
+        } else {
+            pConfig->set(ConfigKey(controllerKey, setting->variableName()), setting->stringify());
+        }
+    }
+}
```

Comment:
This data is going from storage to JavaScript, so wouldn't it be simpler to save JSON files instead of using the config system?

Reply:
We shouldn't clutter the configuration setting to different storage systems. Our config system is the entity, which is supposed to be used for such data.
But migrating the config system to use a modern file format like JSON, might be a good feature on it's own.

Reply:
I also started going that direction at first @Be-ing but got concerned on creating discrepancy with something that, IMHO, should be generic to the whole project. It definitely would be great to use some other formats tho, like JSON or Yaml, but that would make sense to apply for everything I think.

Reply:
FWIW, we already have parts of Mixxx that save data outside the config system, namely the sound hardware XML file and effects state & presets. It's not unprecedented. I don't have a strong opinion about this, but I think it's worth considering.

Reply:
It would be nice if the code was at least flexible enough so we can transition later. If we ever get around to the es6-module-based controller engine, I'd be in favor of ditching the XML for something more human-readable/trendy.

Reply:
Any suggestions on what you mean about flexibility? Happy to look into that, but AFAICT, this should be fairly easy to migrate if we ever want to. Perhaps it could uses QVariant instead of string serialisation, but the reason of going this way is that the definition (in the controller mapping) has to be a string, so I decided to align on that side.

As per discussed in the original issue before starting the work, this feature is here as a step forward, but doesn't intent to be a perfect solution, and since a large refactor on settings is to be expected soon, maybe best not to over complicate the solution?

Reply:
> Any suggestions on what you mean about flexibility?

IIUC don't store it in mixxx.cfg, but in a separate file, preferrably in the user controller directory so custom controller settings are portable?

Reply:
> Any suggestions on what you mean about flexibility? Happy to look into that, but AFAICT, this should be fairly easy to migrate if we ever want to.

I haven't looked at the code too closely (I got hung up on the complications around the memory issues arising from inheritance). It would primarily be nice if the interface was abstract enough to not depend on the underlying serialization format, so it would be easy to read the same data from another file, possibly JSON, YAML, TOML, etc.

Reply:
Are you sure this wouldn't clutter the feature and PR? It is already a chunky bit to review as you suggested, but also making sure it support any kind of object format+ensuring easy integration with the global settings once it gets refactored will add a significant amount of changes.
Just thought I would point out, but at the moment, and because controller settings are set to remain simple atomic types, this is already editable in the .cfg file, and thus exportable as well.
I will try to implement some naive solution to dump or load settings from an object file, and implement that to some degree.

Reply:
I fully agree, that the scope of this PR should be limited, to allow a review!
Changeing of file formats can be done in Follow-Up PRs.